Dashboard creation

pfafflabatuiuc opened the pull request:
First version of the dashboard/logger.

wpfff commented on instrumentserver/dashboard/config.py:
```diff
@@ -0,0 +1,56 @@
+config = {
```
this file looks like an example, right? 
maybe it'd be better make this more clear?
My favorite option would be to have an 'example' folder in the repository root directory.

wpfff commented on instrumentserver/dashboard/dashboard.py:
```diff
@@ -0,0 +1,290 @@
+class PlotParameter:
+    """
+    PlotParameter is used to hold an individual parameter of the plot.
+    It saves the data of it as well as draws the lines for the figures.
```
is it still saving at this moment?

pfafflabatuiuc replied:
It was a typo, I corrected it.

wpfff commented:
@marcosfrenkel -- it looks like we have a bunch of conflicts. it seems that adding a script to be able to launch the parameter manager as command is conflicting with the dashboard stuff. could you have a look and fix that?
The best way would be to merge the changes from the current master branch into your dashboard branch, and then push that branch again.

i'm also adding a few other comments here and there.

wpfff commented on instrumentserver/testing/dummy_instruments/generic.py:
```diff
@@ -38,4 +39,49 @@ def __init__(self, name: str, *args, **kwargs):
\ No newline at end of file
+        return a, b, args[0], c, kwargs['d'], self.param0()
+
+class DummyInstrumentRandomNumber(Instrument):
```
I don't understand this instrument.
how are the random numbers set? do you have to call the generate_data function somewhere?
I would've expected simply that calling the parameter would generate random numbers?

pfafflabatuiuc replied:
I could not make this work by overriding the get function of this instrument. When I call it something else (not get) and hard-code the new name into the logger, the instrument works as intended. However if I leave the logger to use the generic qcodes instrument get function it goes straight into the Instrument get and not the overridden method. I could not figure out how to make it properly work. 

At the moment the logger and dashboard are technically functioning with this instrument, but it is not creating random numbers.

wpfff commented on setup.py:
```diff
@@ -9,5 +9,8 @@
       packages=['instrumentserver'],
       zip_safe=False,
-      entry_points={"console_scripts": ["instrumentserver = instrumentserver.apps:serverScript"]}
+      entry_points={"console_scripts": ["instrumentserver = instrumentserver.apps:serverScript",
```
these script names are very generic and it's not immediately clear that they belong to the instrumentserver package.
I would prefer naming them 'instrumentserver-dashboard', or something like that.

wpfff commented on instrumentserver/dashboard/logger.py:
```diff
@@ -0,0 +1,175 @@
+        # check that the source type is parameter
+        if self.source_type == 'parameter':
+
+            # just for development
```
this should not be in here, right?

wpfff commented on instrumentserver/dashboard/config.py:
```diff
@@ -0,0 +1 @@
+config = {}
```
do we need this file?

wpfff commented on instrumentserver/dashboard/dashboard.py:
```diff
@@ -0,0 +1,281 @@
+from bokeh.palettes import Category10
+
+
+from .__init__ import read_config
```
you don't need the __init__ part. just from . import read_config is fine.

wpfff commented on instrumentserver/dashboard/__init__.py:
```diff
@@ -0,0 +1,257 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The file should be located in the dashboard folder inside of the instrumentserver package and should be called:
+'config.py'. The file should have a python dictionary named config, this is the dictionary that is being read.
```
this is not good. 
the user should never have to put stuff into the package.

a solution could be to look for a config file in the current working directory (with a default name -- say 'instrumentserver-dashboard-cfg.py') and the user should be able to specify a filepath for one, alternatively.

wpfff replied:
you can import from a specific file using importlib.util -- see for instance: https://codeigo.com/python/import-function-from-file#import-file-from-another-location

wpfff commented on instrumentserver/dashboard/logger.py:
```diff
@@ -0,0 +1,175 @@
+    def __init__(self):
+        super().__init__()
+
+        # used for testing, the instruments should be already created for the dashboard to work
```
i'd prefer if the test wasn't hardcoded into this file. a better solution would be to have a separate testing script.

wpfff commented on instrumentserver/dashboard/__init__.py:
```diff
@@ -0,0 +1,257 @@
+Example:
+^^^^^^^^
+
+The following is an example dictionary::
```
i think we should adapt the config structure a bit:
i think it makes more sense to have entries 'plots' and 'options' (i.e., add one more level of nesting for the plots). i think it's too confusing to have one particular name ('options') not be a plot.

pfafflabatuiuc commented:
I have addressed most reviews (check the only unresolved above). I could not get the random number dummy instrument to properly work but decided to stop wasting time on it and work on the rest of the pull request.  

I have the dashboard and logger running and I plan on leaving them to run overnight as the fridge cools down. Right now everything else seems to be working as expected.

pfafflabatuiuc commented:
Just an update about the logger/dasher overnight test. The dashboard and logger run overnight logging 8 temperature channels and 2 pressure channels of the fridge. For all of them it gathered an update once every minute and in about 20 hours the file generated was a little bit less than a 1MB.
